src/Documentation/clusters_and_clients/index.md

```diff
@@ -86,6 +86,117 @@ So it is a responsibility of the application code to build a higher level reliab
+If `Connect` returns successfully, the cluster client is guaranteed to be usable until it is disposed. This means that even if the client experiences connection issues, it will attempt to recover indefinitely. The exact recovery behavior can be configured on a `GatewayOptions` object provided by the `ClientBuilder`, e.g.:
+
+```csharp
+var client = new ClientBuilder()
+    // ...
+    .Configure<GatewayOptions>(opts => GatewayListRefreshPeriod = TimeSpan.FromMinutes(10)) // Default is 1 min.
+    .Build();
+```
+
+In the second case, where a connection issue occurs during a grain call, a `SiloUnavailableException` will be thrown on the client side. This could be handled like so:
+
+``` csharp
+IPlayerGrain player = client.GetGrain<IPlayerGrain>(playerId);
+
+try
+{
+    await player.JoinGame(game);
+}
+catch (SiloUnavailableException)
+{
+    // Lost connection to the cluster...
+}
+```
+
+The grain reference is not invalidated in this situation; the call could be retried on the same reference later, when a connection might have been re-established.
+
+### Dependency Injection
+
+The recommended way to use a cluster client in a program that uses the .NET Generic Host is to inject an `IClusterClient` singleton instance via dependency injection, which can then be accepted as a constructor parameter in hosted services, ASP.NET controllers, etc.
+
+Note that when co-hosting an Orleans silo in the same process that will be connecting to it, it is *not* necessary to manually create a client; Orleans will automatically provide one and manage its lifetime appropriately.
+
+When connecting to a cluster in a different process (e.g. on a different machine), a common pattern is to create a hosted service like this:
+
+```csharp
+public class ClusterClientHostedService : IHostedService
+{
+    public IClusterClient Client { get; }
+
+    public ClusterClientHostedService(ILoggerProvider loggerProvider)
+    {
+        Client = new ClientBuilder()
+            // Appropriate client configuration here, e.g.:
+            .UseLocalhostClustering()
+            .ConfigureLogging(builder => builder.AddProvider(loggerProvider))
+            .Build();
+    }
+
+    public async Task StartAsync(CancellationToken cancellationToken)
+    {
+        // A retry filter could be provided here.
+        await Client.Connect();
+    }
+
+    public async Task StopAsync(CancellationToken cancellationToken)
+    {
+        await Client.Close();
+
+        Client.Dispose();
+    }
+}
+```
+
+The service is then registered like this:
+
+```csharp
+public class Program
+{
+    static Task Main()
+    {
+        return new HostBuilder()
+            .ConfigureServices(services =>
+            {
+                services.AddSingleton<ClusterClientHostedService>();
+                services.AddSingleton<IHostedService>(sp => sp.GetService<ClusterClientHostedService>());
+                services.AddSingleton<IClusterClient>(sp => sp.GetService<ClusterClientHostedService>().Client);
```

Comment:
We should add IGrainFactory here, too

```suggestion
                services.AddSingleton<IClusterClient>(sp => sp.GetService<ClusterClientHostedService>().Client);
                services.AddSingleton<IGrainFactory>(sp => sp.GetService<ClusterClientHostedService>().Client);
```

Reply:
This does make me wonder when you would actually want to use `IGrainFactory` specifically. `IClusterClient` exposes more functionality and seems to always be available when an `IGrainFactory` is, so is there actually a reason one might want to get an `IGrainFactory` specifically?

No objection to the suggested change either way, though.

Reply:
Yes, `IGrainFactory` is usually the preferred choice. The main difference, as far as application developers are concerned, is that `IClusterClient` has a convenience method for getting stream providers.